File: examples/xls_controller.py
import threading
import numpy as np
from pynput import keyboard     # Global keyboard listener
from typing import Dict
from dyros_robot_controller import KinematicParam, DriveType
from dyros_robot_controller.robot_data.mobile.base import MobileBase as RobotData
from dyros_robot_controller.robot_controller.mobile.base import MobileControllerBase as RobotController

class XLSController:
    def __init__(self, dt: float):
        # Simulation/control timestep
        self.dt = dt
        
        # Define mobile base kinematics (Mecanum drive) and constraints
        # All geometry is expressed in the base frame:
        #  - base2wheel_positions: 2D position of each wheel center wrt base origin
        #  - base2wheel_angles: wheel steering angles (fixed here)
        #  - roller_angles: Mecanum roller angles for each wheel
        #  - max_*: saturation limits (used by controller for velocity/acceleration limits)
        param = KinematicParam(
            type                 = DriveType.Mecanum,
            wheel_radius         = 0.120,
            base2wheel_positions = [np.array([ 0.2225,  0.2045]),   # Front Left
                                    np.array([ 0.2225, -0.2045]),   # Front Right
                                    np.array([-0.2225,  0.2045]),   # Rear Left
                                    np.array([-0.2225, -0.2045])],  # Rear Right
            base2wheel_angles    = [0,0,0,0],
            roller_angles        = [-np.pi/4,  # Front Left
                                     np.pi/4,  # Front Right
                                     np.pi/4,  # Rear Left
                                    -np.pi/4], # Rear Right 
            max_lin_speed        = 2,
            max_ang_speed        = 3,
            max_lin_acc          = 3,
            max_ang_acc          = 6,
        )

        # Initialize Dyros mobile robot model and controller
        self.robot_data = RobotData(param)
        self.robot_controller = RobotController(dt=self.dt,
                                                robot_data=self.robot_data)
        
        # Number of wheels (derived from param)
        self.wheel_num = self.robot_data.get_wheel_num()
        
        # Wheel states and command buffers
        self.wheel_pos           = np.zeros(self.wheel_num)  # measured wheel positions
        self.wheel_vel           = np.zeros(self.wheel_num)  # measured wheel velocities
        self.wheel_vel_desired   = np.zeros(self.wheel_num)  # desired wheel velocities (control input)
        
        # Desired base twist in base frame: [vx, vy, w]
        self.base_vel_desired    = np.zeros(3)

        # --- Control mode bookkeeping ---
        self.control_mode = "Stop"         # Default: stop (zero command)
        self.is_control_mode_changed = True
        self.sim_time = 0.0
        self.control_start_time = 0.0
        
        # Start a global keyboard listener for non-blocking input
        # We track currently pressed keys to allow multi-key combos (e.g., Up+Right)
        self._pressed_keys = set()
        self._keys_lock = threading.Lock() 
        self._listener = keyboard.Listener(on_press=self._on_key_press, on_release=self._on_key_release)
        self._listener.daemon = True
        self._listener.start()


    def update_model(self, current_time: float, qpos_dict: Dict, qvel_dict: Dict):
        # Update simulation time from MuJoCo
        self.sim_time = current_time

        # Map sim joint dictionaries -> ordered wheel arrays
        # NOTE: The order must match the order assumed by KinematicParam above.
        self.wheel_pos[:] = np.array([qpos_dict["front_left_wheel"],   # Front Left
                                      qpos_dict["front_right_wheel"],  # Front Right
                                      qpos_dict["rear_left_wheel"],    # Rear Left
                                      qpos_dict["rear_right_wheel"]])  # Rear Right
        self.wheel_vel[:] = np.array([qvel_dict["front_left_wheel"],
                                      qvel_dict["front_right_wheel"],
                                      qvel_dict["rear_left_wheel"],
                                      qvel_dict["rear_right_wheel"]])
        
        # Push latest wheel states into the model
        self.robot_data.update_state(self.wheel_pos, self.wheel_vel)

    def compute(self):
        # One-time initialization per mode entry
        if self.is_control_mode_changed:
            self.is_control_mode_changed = False
            self.control_start_time = self.sim_time
            
        # --- Mode: Stop (zero command) ---
        if self.control_mode == "Stop":
            self.wheel_vel_desired[:] = 0.0
            
        # --- Mode: Base Velocity Tracking ---
        # Convert current key set -> desired base twist [vx, vy, w]
        # Then compute wheel velocities via the controller's mapping
        elif self.control_mode == "Base Velocity Tracking":
            self.base_vel_desired = self._keys_to_xdot()
            self.wheel_vel_desired = self.robot_controller.velocity_command(self.base_vel_desired)
      
        # Format controller outputs for MuJoCo (actuator name -> value)
        return {
            "front_left_wheel":  float(self.wheel_vel_desired[0]),
            "front_right_wheel": float(self.wheel_vel_desired[1]),
            "rear_left_wheel":  float(self.wheel_vel_desired[2]),
            "rear_right_wheel": float(self.wheel_vel_desired[3]),
        }
    
    def set_mode(self, control_mode: str):
        # Switch control mode and trigger per-mode re-initialization in compute()
        self.is_control_mode_changed = True
        self.control_mode = control_mode
        print(f"Control Mode Changed: {self.control_mode}")
    
        
    def _on_key_press(self, key):
        # Track “held” keys to support continuous velocity while pressed
        with self._keys_lock:
            self._pressed_keys.add(key)

        # Numeric hotkeys for mode switching
        try:
            if key.char == '1':
                self.set_mode("Stop")
            elif key.char == '2':
                self.set_mode("Base Velocity Tracking")
            elif key.char == '3':
                self.set_mode("Gravity Compensation")  # (kept for parity; not used in compute)
        except AttributeError:
            # Special keys (arrows, etc.) don’t have .char
            pass

    def _on_key_release(self, key):
        # Remove key from the "held" set → that axis goes to 0 next cycle
        with self._keys_lock:
            if key in self._pressed_keys:
                self._pressed_keys.remove(key)
                
    def _keys_to_xdot(self) -> np.ndarray:
        """Map currently held keys to a base twist [vx, vy, w] in the base frame.
        - Arrow keys control linear vx, vy
        - 'b' increases yaw rate (+w), 'v' decreases yaw rate (-w)
        Releasing a key removes it from the set, so the corresponding component becomes 0.
        """
        vx = vy = w = 0.0
        with self._keys_lock:
            held = set(self._pressed_keys)

        # Linear velocity with arrow keys
        if keyboard.Key.up in held:    vx += 1.0   # forward
        if keyboard.Key.down in held:  vx -= 1.0   # backward
        if keyboard.Key.right in held: vy += 1.0   # left-to-right strafe
        if keyboard.Key.left in held:  vy -= 1.0   # right-to-left strafe

        # Yaw rate with letters ('b' = +w, 'v' = -w)
        if 'b' in [getattr(k, 'char', None) for k in held]:
            w += 1.0
        if 'v' in [getattr(k, 'char', None) for k in held]:
            w -= 1.0

        # Return desired base twist; controller will map to wheel velocities
        return np.array([vx, vy, w], dtype=float)

File: examples/C++/src/xls_controller.cpp
#include "xls_controller.hpp"

#ifndef ROBOTS_DIRECTORY
#  error "ROBOTS_DIRECTORY is not defined. Define in CMake: -DROBOTS_DIRECTORY=\"${CMAKE_SOURCE_DIR}/../robots\""
#endif

XLSController::XLSController(const double dt)
: dt_(dt)
{
  // Define mobile base kinematics (Mecanum drive) and constraints
  // All geometry is expressed in the base frame:
  //  - base2wheel_positions: 2D position of each wheel center wrt base origin
  //  - base2wheelㄴ_angles: wheel steering angles (fixed here)
  //  - roller_angles: Mecanum roller angles for each wheel
  //  - max_*: saturation limits (used by controller for velocity/acceleration limits)
  drc::Mobile::KinematicParam param;
  param.type = drc::Mobile::DriveType::Mecanum;
  param.wheel_radius = 0.120;
  param.base2wheel_positions = {Eigen::Vector2d( 0.2225,  0.2045),  // Front Left
                                Eigen::Vector2d( 0.2225, -0.2045),  // Front Right
                                Eigen::Vector2d(-0.2225,  0.2045),  // Rear Left
                                Eigen::Vector2d(-0.2225, -0.2045)}; // Rear Right
  param.base2wheel_angles = {0, 0, 0, 0};
  param.roller_angles = {-M_PI/4,  // Front Left
                          M_PI/4,  // Front Right
                          M_PI/4,  // Rear Left
                         -M_PI/4}; // Rear Right
  param.max_lin_speed = 2.0;
  param.max_ang_speed = 3.0;
  param.max_lin_acc = 3.0;
  param.max_ang_acc = 6.0;

  // Initialize robot model and controller (Dyros Robot Controller)
  robot_data_ = std::make_shared<drc::Mobile::RobotData>(param);
  robot_controller_ = std::make_shared<drc::Mobile::RobotController>(dt_, robot_data_);

  // Number of wheels (derived from param)
  wheel_num_ = robot_data_->getWheelNum();

  // Wheel states and command buffers
  wheel_pos_.setZero(wheel_num_);
  wheel_vel_.setZero(wheel_num_);
  wheel_vel_desired_.setZero(wheel_num_);

  // Desired base twist in base frame: [vx, vy, w]
  base_vel_desired_.setZero(3);

  // Information for xls robot
  std::cout << "info: \n" << robot_data_->getVerbose() << std::endl; 

  // Start a keyboard listener (runs in a background thread)
  startKeyListener_();
}

XLSController::~XLSController() 
{
  stopKeyListener_();
}

void XLSController::updateModel(const double current_time,
                                const std::unordered_map<std::string, double>& qpos_dict,
                                const std::unordered_map<std::string, double>& qvel_dict)
{
  // Update simulation time
  sim_time_ = current_time;

  // Read joint positions/velocities from simulation dictionaries
  // NOTE: The order must match the order assumed by KinematicParam above.
  wheel_pos_(0) = qpos_dict.at("front_left_wheel");  // Front Left
  wheel_pos_(1) = qpos_dict.at("front_right_wheel"); // Front Right
  wheel_pos_(2) = qpos_dict.at("rear_left_wheel");   // Rear Left
  wheel_pos_(3) = qpos_dict.at("rear_right_wheel");  // Rear Right
  wheel_vel_(0) = qvel_dict.at("front_left_wheel");
  wheel_vel_(1) = qvel_dict.at("front_right_wheel");
  wheel_vel_(2) = qvel_dict.at("rear_left_wheel");
  wheel_vel_(3) = qvel_dict.at("rear_right_wheel");

  // Push latest wheel states into the model
  robot_data_->updateState(wheel_pos_, wheel_vel_);
}

std::unordered_map<std::string, double> XLSController::compute() 
{
  // One-time initialization per mode entry
  if (is_mode_changed_) 
  {
    is_mode_changed_ = false;
    control_start_time_ = sim_time_;
  }
  // --- Control mode: Stop (zero command) ---
  if (control_mode_ == "Stop") {
    wheel_vel_desired_.setZero();
  }
  // --- Control mode: Base Velocity Tracking ---
  // Convert current key set -> desired base twist [vx, vy, w]
  // Then compute wheel velocities via the controller's mapping
  else if (control_mode_ == "Base Velocity Tracking") 
  {
    base_vel_desired_ = keysToxdot();
    wheel_vel_desired_ = robot_controller_->VelocityCommand(base_vel_desired_);
  }
  
  // Format controller outputs for MuJoCo (actuator name -> value)
  std::unordered_map<std::string, double> ctrl_dict;
  ctrl_dict.reserve(wheel_num_);
  ctrl_dict["front_left_wheel"]  = wheel_vel_desired_(0);
  ctrl_dict["front_right_wheel"] = wheel_vel_desired_(1);
  ctrl_dict["rear_left_wheel"]   = wheel_vel_desired_(2);
  ctrl_dict["rear_right_wheel"]  = wheel_vel_desired_(3);
  return ctrl_dict;
}

void XLSController::setMode(const std::string& control_mode) 
{
  // Change mode and flag reset
  is_mode_changed_ = true;
  control_mode_ = control_mode;
  std::cout << "Control Mode Changed: " << control_mode_ << std::endl;
}

void XLSController::startKeyListener_() 
{
  tty_ok_ = ::isatty(STDIN_FILENO);
  if (!tty_ok_) {
    std::cout << "[XLSController] stdin is not a TTY; keyboard control disabled.\n";
    return;
  }
  setRawMode_();
  stop_key_ = false;
  key_thread_ = std::thread(&XLSController::keyLoop_, this);

  std::cout << "[XLSController] Keyboard: [1]=Stop, [2]=Base Velocity Tracking, arrows=xy, b/v=yaw, [q]=quit listener\n";
}

void XLSController::stopKeyListener_() 
{
  if (!tty_ok_) return;
  stop_key_ = true;
  if (key_thread_.joinable()) key_thread_.join();
  restoreTerm_();
}

void XLSController::keyLoop_() 
{
  using clock = std::chrono::steady_clock;
  const auto clear_after = std::chrono::milliseconds(120);

  while (!stop_key_) {
    fd_set rfds;
    FD_ZERO(&rfds);
    FD_SET(STDIN_FILENO, &rfds);

    struct timeval tv;
    tv.tv_sec = 0;
    tv.tv_usec = 10 * 1000;

    int ret = ::select(STDIN_FILENO + 1, &rfds, nullptr, nullptr, &tv);
    const auto now = clock::now();

    if (ret > 0 && FD_ISSET(STDIN_FILENO, &rfds)) {
      char buf[32];
      ssize_t nread = ::read(STDIN_FILENO, buf, sizeof(buf));
      if (nread > 0) {
        for (ssize_t i = 0; i < nread; ++i) {
          char c = buf[i];
          if (c == 'q' || c == 'Q') { stop_key_ = true; break; }
          else if (c == '1') { setMode("Stop"); }
          else if (c == '2') { setMode("Base Velocity Tracking"); }
          else if (c == '\x1b') {
            if (i + 2 < nread && buf[i+1] == '[') {
              char code = buf[i+2];
              std::lock_guard<std::mutex> lk(keys_mtx_);
              if      (code == 'A') keys_.up    = true;
              else if (code == 'B') keys_.down  = true;
              else if (code == 'C') keys_.right = true;
              else if (code == 'D') keys_.left  = true;
              keys_.last_event = now;
              i += 2;
            }
          } else {
            std::lock_guard<std::mutex> lk(keys_mtx_);
            if (c == 'b' || c == 'B') keys_.b = true;
            if (c == 'v' || c == 'V') keys_.v = true;
            keys_.last_event = now;
          }
        }
      }
    } else {
      std::lock_guard<std::mutex> lk(keys_mtx_);
      if (now - keys_.last_event > clear_after) {
        keys_.up = keys_.down = keys_.left = keys_.right = false;
        keys_.b = keys_.v = false;
      }
    }
  }
}


void XLSController::setRawMode_() 
{
  if (!tty_ok_) return;
  struct termios raw;
  if (tcgetattr(STDIN_FILENO, &orig_term_) == -1) 
  {
    perror("tcgetattr");
    tty_ok_ = false;
    return;
  }
  raw = orig_term_;
  raw.c_lflag &= ~(ICANON | ECHO);
  raw.c_cc[VMIN]  = 0;
  raw.c_cc[VTIME] = 1;
  if (tcsetattr(STDIN_FILENO, TCSANOW, &raw) == -1) 
  {
    perror("tcsetattr");
    tty_ok_ = false;
  }
}

void XLSController::restoreTerm_() 
{
  if (!tty_ok_) return;
  if (tcsetattr(STDIN_FILENO, TCSANOW, &orig_term_) == -1) 
  {
    perror("tcsetattr restore");
  }
}

Eigen::Vector3d XLSController::keysToxdot() const
{
  double vx = 0.0, vy = 0.0, w = 0.0;
  {
    std::lock_guard<std::mutex> lk(keys_mtx_);
    if (keys_.up)    vx += 1.0;   // forward
    if (keys_.down)  vx -= 1.0;   // backward
    if (keys_.left)  vy += 1.0;   // left-to-right
    if (keys_.right) vy -= 1.0;   // right-to-left
    if (keys_.b)     w  += 1.0;   // yaw +
    if (keys_.v)     w  -= 1.0;   // yaw -
  }
  return Eigen::Vector3d(vx, vy, w);
}